va/validation-authority.go

```diff
@@ -330,18 +282,142 @@ func (va *ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdenti
 		}
 		err = challenge.Error
-		return challenge, err
+		return emptyBody, challenge, err
 	}
 
 	// Read body & test
-	body, readErr := ioutil.ReadAll(httpResponse.Body)
-	if readErr != nil {
+	body, err := ioutil.ReadAll(httpResponse.Body)
+	if err != nil {
+		challenge.Status = core.StatusInvalid
+		challenge.Error = &core.ProblemDetails{
+			Type:   core.UnauthorizedProblem,
+			Detail: fmt.Sprintf("Error reading HTTP response body: %v", err),
+		}
+		return emptyBody, challenge, err
+	}
+
+	return body, challenge, nil
+}
+
+func (va *ValidationAuthorityImpl) validateZName(identifier core.AcmeIdentifier, input core.Challenge, zName string) (core.Challenge, error) {
+	challenge := input
+
+	addr, allAddrs, problem := va.getAddr(identifier.Value)
+	challenge.ValidationRecord = []core.ValidationRecord{
+		core.ValidationRecord{
+			Hostname:          identifier.Value,
+			AddressesResolved: allAddrs,
+			AddressUsed:       addr,
+		},
+	}
+	if problem != nil {
+		challenge.Status = core.StatusInvalid
+		challenge.Error = problem
+		return challenge, challenge.Error
+	}
+
+	// Make a connection with SNI = nonceName
+	portString := fmt.Sprintf("%d", va.tlsPort)
+	hostPort := net.JoinHostPort(addr.String(), portString)
+	challenge.ValidationRecord[0].Port = portString
+	va.log.Notice(fmt.Sprintf("%s [%s] Attempting to validate for %s %s", challenge.Type, identifier, hostPort, zName))
+	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: validationTimeout}, "tcp", hostPort, &tls.Config{
+		ServerName:         zName,
+		InsecureSkipVerify: true,
+	})
+
+	if err != nil {
 		challenge.Status = core.StatusInvalid
 		challenge.Error = &core.ProblemDetails{
+			Type:   parseHTTPConnError(err),
+			Detail: "Failed to connect to host for DVSNI challenge",
+		}
+		va.log.Debug(fmt.Sprintf("%s [%s] TLS Connection failure: %s", challenge.Type, identifier, err))
+		return challenge, err
+	}
+	defer conn.Close()
+
+	// Check that zName is a dNSName SAN in the server's certificate
+	certs := conn.ConnectionState().PeerCertificates
+	if len(certs) == 0 {
+		challenge.Error = &core.ProblemDetails{
 			Type:   core.UnauthorizedProblem,
-			Detail: fmt.Sprintf("Error reading HTTP response body"),
+			Detail: "No certs presented for TLS SNI challenge",
+		}
+		challenge.Status = core.StatusInvalid
+		return challenge, challenge.Error
+	}
+	for _, name := range certs[0].DNSNames {
+		if subtle.ConstantTimeCompare([]byte(name), []byte(zName)) == 1 {
+			challenge.Status = core.StatusValid
+			return challenge, nil
+		}
+	}
+
+	challenge.Error = &core.ProblemDetails{
+		Type:   core.UnauthorizedProblem,
+		Detail: "Correct zName not found for TLS SNI challenge",
+	}
+	challenge.Status = core.StatusInvalid
+	return challenge, challenge.Error
+}
+
+//-----BEGIN TO DELETE-----
+func verifyValidationJWS(validation *jose.JsonWebSignature, accountKey *jose.JsonWebKey, target map[string]interface{}) error {
```

Comment:
It will clean up the diff if this were moved back up top. Just add separate "TO DELETE" fencing for it.